src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/HeaderModelBinder.cs

```diff
@@ -45,86 +45,83 @@ public HeaderModelBinder(ILoggerFactory loggerFactory)
             _logger = loggerFactory.CreateLogger<HeaderModelBinder>();
         }
 
+        /// <summary>
+        /// Initializes a new instance of <see cref="HeaderModelBinder"/>.
+        /// </summary>
+        /// <param name="innerModelBinder">The <see cref="IModelBinder"/> which does the actual
+        /// binding of values.</param>
+        /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
+        public HeaderModelBinder(ILoggerFactory loggerFactory, IModelBinder innerModelBinder)
+        {
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            if (innerModelBinder == null)
+            {
+                throw new ArgumentNullException(nameof(innerModelBinder));
+            }
+
+            InnerModelBinder = innerModelBinder;
+            _logger = loggerFactory.CreateLogger<HeaderModelBinder>();
+        }
+
+        // to enable unit testing
+        internal IModelBinder InnerModelBinder { get; }
+
         /// <inheritdoc />
-        public Task BindModelAsync(ModelBindingContext bindingContext)
+        public async Task BindModelAsync(ModelBindingContext bindingContext)
         {
             if (bindingContext == null)
             {
                 throw new ArgumentNullException(nameof(bindingContext));
             }
 
-            var request = bindingContext.HttpContext.Request;
+            _logger.AttemptingToBindModel(bindingContext);
 
             // Property name can be null if the model metadata represents a type (rather than a property or parameter).
             var headerName = bindingContext.FieldName;
 
-            _logger.AttemptingToBindModel(bindingContext);
-
+            var request = bindingContext.HttpContext.Request;
             if (!request.Headers.ContainsKey(headerName))
             {
                 _logger.FoundNoValueInRequest(bindingContext);
             }
 
-            object model;
-            if (bindingContext.ModelType == typeof(string))
-            {
-                var value = request.Headers[headerName];
-                model = (string)value;
-            }
-            else if (ModelBindingHelper.CanGetCompatibleCollection<string>(bindingContext))
+            // Do not set ModelBindingResult to Failed on not finding the value as we want the inner modelbinder to
+            // do that. This would give a chance to the inner binder to add more useful information. For example,
+            // SimpleTypeModelBinder adds a model error when binding to let's say and integer and the model is null.
+
+            // Explicitly pass in the header name as the key rather than taking in the model name to look for values
+            // as otherwise it would be breaking from earlier version where we didn't consider prefixes.
```

Comment:
I would explain this a bit differently. Using the ModelName directly rather than the model path (full name) isn't just a back-compat concern. It's always the way this is supposed to work. There was never a desire for model binding a header value to compose keys like we do for forms and query strings.

Reply:
> Explicitly pass in the header name 

Where is the code that does this? I would expect to see this show up at L113 or so